library/common/extensions/filters/http/assertion/filter.cc

```diff
@@ -26,53 +28,91 @@ AssertionFilter::AssertionFilter(AssertionFilterConfigSharedPtr config) : config
 Http::FilterHeadersStatus AssertionFilter::decodeHeaders(Http::RequestHeaderMap& headers,
                                                          bool end_stream) {
   config_->rootMatcher().onHttpRequestHeaders(headers, statuses_);
-  if (!config_->rootMatcher().matchStatus(statuses_).matches_) {
+  auto& matchStatus = config_->rootMatcher().matchStatus(statuses_);
+  if (!matchStatus.matches_ && !matchStatus.might_change_status_) {
     decoder_callbacks_->sendLocalReply(Http::Code::BadRequest,
                                        "Request Headers do not match configured expectations",
                                        nullptr, absl::nullopt, "");
     return Http::FilterHeadersStatus::StopIteration;
   }
 
   if (end_stream) {
-    decoder_callbacks_->sendLocalReply(Http::Code::OK,
-                                       "Request Headers match configured expectations", nullptr,
-                                       absl::nullopt, "");
-    return Http::FilterHeadersStatus::StopIteration;
+    Buffer::OwnedImpl empty_buffer;
```

Comment:
I see, so here we would like to see if there were expectations for body and trailers that because this request ended on a headers frame would not match.

Do you mind adding comments explaining the thought process for posterity?